### ISSUE TITLE
Host with multiple transports

### ISSUE THREAD

**Issue by esso23** [Code-Sharp/WampSharp#328]
Hello,

I need to create a Host that has multiple transports having different types of authentication.
My case:
- Transport at port 50000 using secure websocket with OAuth authentication (for public clients) - security
- Transport at port 50001 using RawSocket transport using ticket authentication (for local clients) - speed

I tried using code from [WampAuthenticationHost](https://github.com/Code-Sharp/WampSharp/blob/master/src/netstandard/WampSharp/WAMP2/V2/Authentication/Host/WampAuthenticationHost.cs) but I ended up with endless internal dependency chain (I gave up after copying more than dozen classes).

I just need to have 2 different `RegisterTransport` methods, one which uses `CreateAuthenticationBinding` (this one uses `WampAuthenticationBinaryBinding`, `WampAuthenticationBinaryBinding` and `WampAuthenticationBinding` which are all internal and have other internal dependencies) and one which does not.

I also tried to use implementation of `WampHost` and edit that but that's even worse, almost everything is internal.

Is there a simple way to do this? Or is it at least possible to make most of those things public?

Thanks
Esso

**Comment by darkl** [comment]
It seems to me like the problem is that you have a single `IWampSessionAuthenticatorFactory` so apparently you can't use two authentication methods?
But `IWampSessionAuthenticatorFactory` actually receives as an argument of the type `WampPendingClientDetails` so in theory we can just add the details about the underlying transport there and then you can choose your authenticator factory based on that?

Elad

**Comment by esso23** [comment]
Oh I didn't know it's sent in the arguments like that. That's very convenient. I actually have two separatate `IWampSessionAuthenticatorFactory`s so I created one Master factory which has a dictionary to switch the factory based on transport.
Is there some list where all the transport strings are listed?

**Comment by darkl** [comment]
Can you attach here your master factory code please?
For future reference and for understanding what you exactly did.

Thanks
Elad

**Comment by esso23** [comment]
Well it does not work since the `TransportDetails` object is null.


![Screenshot 2020-12-10 22 58 16](https://user-images.githubusercontent.com/5548662/101834975-5def3880-3b3b-11eb-85b1-23ab65f69f2d.png)

**Comment by darkl** [comment]
Yeah, I'll fix that. Maybe the WampPendingClientDetails class should also
include the transport itself for the case where you have two instances of a
transport of the same type with different authorization methods.

Elad

**Comment by esso23** [comment]
I was thinking about it and in my opinion it's a hacky solution.
Iideally we should be able to just attach `IWampSessionAuthenticatorFactory` to a binding just like it's done in `WampAuthenticationHost` and that would solve all the issues.
In that case I would be able to not only have different authentication methods for multiple instances of the same transport, but even specify different authenticators for each binding on the same transport.

**Comment by darkl** [comment]
We'll have to change some api.

```csharp
// Allows to map different bindings to different authentication methods
public void RegisterTransport(IWampTransport transport, IDictonary<IWampBinding, IWampSessionAuthenticatorFactory> bindingToAuthenticationFactory);
```

If we remove the original overload:
```csharp
public void RegisterTransport(IWampTransport transport, IEnumerable<IWampBinding> bindings)
```
this will break api and also won't allow `WampAuthenticationHost` to implement `IWampHost`. Another option is just to add a new class instead of `WampAuthenticationHost` that has this functionality. I'll have to do some refactoring tricks but this is doable.

I also checked and authentication for `crossbario` is also defined per transport so we should probably support this.

**Comment by darkl** [comment]
Hi Esso,

I just push a commit with this feature. Can you try it out? The version is `21.0.0-develop-33` from GitHub and the new class is called `WampMultiAuthenticationHost`.

Usage:
```csharp
WampMultiAuthenticationHost host =
    new WampMultiAuthenticationHost();

host.RegisterTransport(new FleckWebSocketTransport("ws://127.0.0.1:8080/ws"),
                       new Dictionary<IWampBinding, IWampSessionAuthenticatorFactory>()
                       {
                           {new JTokenJsonBinding(), new WampCraUserDbAuthenticationFactory(new MyAuthenticationProvider(), new MyUserDb())},
                           {new JTokenMessagePackBinding(), new MyAuthenticatorFactory()}
                       });
```

Elad

**Comment by esso23** [comment]
Hi Elad,
sorry for answering so late. I tried this (using `21.0.0-develop-34`) and it seems to work without any issues (I will do more testing in the following weeks).

I propose adding another overload:
```cs
void RegisterTransport(IWampTransport transport, IEnumerable<IWampBinding> bindings, IWampSessionAuthenticatorFactory sessionAuthenticatorFactory);
```

Since it's likely going to be the most common use-case to add the same auth factory to all the bindings on the given transport.

**Comment by esso23** [comment]
Hi @darkl, 
any ETA on when we can expect this to be merged to master?

Thanks,
Esso

**Comment by darkl** [comment]
Hi Esso,

I'm currently very busy. I'll have more time after March 15th.

Elad

**Comment by esso23** [comment]
Alright, we'll be patiently waiting then :)
